File: src/table.ts
import type { WorktreeView } from "./types";
import { formatStatus } from "./status";

const HEADERS = ["", "BRANCH", "PATH", "COMMIT", "STATUS"] as const;

function width(value: string): number {
  return Bun.stringWidth(value.replace(/\x1b\[[0-9;]*m/g, ""));
}

export function truncate(value: string, maxWidth: number): string {
  if (maxWidth <= 0) return "";
  if (width(value) <= maxWidth) return value;
  if (maxWidth === 1) return "…";
  let result = "";
  for (const character of value) {
    if (width(result + character) > maxWidth - 1) break;
    result += character;
  }
  return result + "…";
}

function pad(value: string, target: number): string {
  return value + " ".repeat(Math.max(0, target - width(value)));
}

function rows(worktrees: WorktreeView[]): string[][] {
  return worktrees.map((worktree) => [
    worktree.current ? "*" : "",
    worktree.branch ?? "detached",
    worktree.path,
    worktree.head ? worktree.head.slice(0, 12) : "-",
    formatStatus(worktree),
  ]);
}

export function renderTable(worktrees: WorktreeView[], terminalWidth = 120): string[] {
  const data = rows(worktrees);
  const allRows = [Array.from(HEADERS), ...data];
  const natural = HEADERS.map((_, column) => Math.max(...allRows.map((row) => width(row[column] ?? ""))));
  const gaps = (HEADERS.length - 1) * 2;
  const minimums = [1, 3, 3, 4, 3];
  const widths = [...natural];
  while (widths.reduce((sum, item) => sum + item, gaps) > terminalWidth) {
    let candidate = -1;
    let excess = 0;
    for (let column = 0; column < widths.length; column++) {
      const room = widths[column]! - minimums[column]!;
      if (room > excess) { candidate = column; excess = room; }
    }
    if (candidate === -1) break;
    widths[candidate]!--;
  }
  const format = (row: readonly string[]) => row.map((cell, column) => {
    const clipped = truncate(cell, widths[column]!);
    return column === row.length - 1 ? clipped : pad(clipped, widths[column]!);
  }).join("  ");
  return [format(HEADERS), ...data.map(format)];
}

File: src/table.rs
use crate::{status, theme::Theme, types::Worktree};
use unicode_width::{UnicodeWidthChar, UnicodeWidthStr};
const HEADERS: [&str; 5] = ["", "BRANCH", "PATH", "COMMIT", "STATUS"];
fn width(s: &str) -> usize {
    UnicodeWidthStr::width(s)
}
pub fn truncate(value: &str, max: usize) -> String {
    if max == 0 {
        return String::new();
    }
    if width(value) <= max {
        return value.into();
    }
    if max == 1 {
        return "…".into();
    }
    let mut out = String::new();
    let mut used = 0;
    for c in value.chars() {
        let w = c.width().unwrap_or(0);
        if used + w > max - 1 {
            break;
        }
        out.push(c);
        used += w
    }
    out.push('…');
    out
}
fn pad(value: &str, target: usize) -> String {
    format!(
        "{}{}",
        value,
        " ".repeat(target.saturating_sub(width(value)))
    )
}
pub fn render(
    worktrees: &[Worktree],
    terminal_width: usize,
    color: bool,
    selected: Option<usize>,
) -> Vec<String> {
    let rows: Vec<[String; 5]> = worktrees
        .iter()
        .map(|w| {
            [
                if w.current { "*" } else { "" }.into(),
                w.branch.clone().unwrap_or_else(|| "detached".into()),
                w.path.clone(),
                if w.head.is_empty() {
                    "-".into()
                } else {
                    w.head.chars().take(12).collect()
                },
                status::format(w),
            ]
        })
        .collect();
    let mut widths = [0; 5];
    for (i, h) in HEADERS.iter().enumerate() {
        widths[i] = width(h)
    }
    for row in &rows {
        for i in 0..5 {
            widths[i] = widths[i].max(width(&row[i]))
        }
    }
    let mins = [1, 3, 3, 4, 3];
    while widths.iter().sum::<usize>() + 8 > terminal_width {
        let Some((i, _)) = widths
            .iter()
            .enumerate()
            .filter(|(i, w)| **w > mins[*i])
            .max_by_key(|(i, w)| **w - mins[*i])
        else {
            break;
        };
        widths[i] -= 1
    }
    let format_row = |r: &[String; 5]| -> Vec<String> {
        (0..5)
            .map(|i| {
                let s = truncate(&r[i], widths[i]);
                if i == 4 {
                    s
                } else {
                    pad(&s, widths[i])
                }
            })
            .collect()
    };
    let theme = Theme { color };
    let hs = HEADERS.map(String::from);
    let header = format_row(&hs)
        .iter()
        .map(|s| theme.header(s))
        .collect::<Vec<_>>()
        .join("  ");
    let mut out = vec![header];
    for (index, row) in rows.iter().enumerate() {
        let cells = format_row(row);
        let w = &worktrees[index];
        out.push(
            cells
                .iter()
                .enumerate()
                .map(|(col, s)| match col {
                    0 if w.current => theme.current(s),
                    1 if selected == Some(index) => theme.selected(s),
                    1 if w.branch.is_some() => theme.branch(s),
                    1 => theme.detached(s),
                    2 if selected == Some(index) => theme.wrap_selected_path(s),
                    2 => s.clone(),
                    3 => theme.commit(s),
                    4 => theme.status(w),
                    _ => s.clone(),
                })
                .collect::<Vec<_>>()
                .join("  "),
        )
    }
    out
}
